src/sinks/file2.rs

```diff
@@ -0,0 +1,197 @@
+use crate::{
+    template::Template,
+    topology::config::{DataType, SinkConfig, SinkContext},
+    Event, Result,
+};
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::{Poll, Sink, StartSend};
+use futures03::channel::mpsc::{channel, Receiver, Sender};
+use futures03::compat::CompatSink;
+use futures03::StreamExt;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::{ffi, path};
+use tokio02::{fs::File, io::AsyncWriteExt};
+
+// === StreamingSink ===
+
+#[async_trait]
+pub trait StreamingSink: Send + Sync + 'static {
+    async fn run(&mut self, input: Receiver<Event>) -> Result<()>;
+
+    fn build_sink(self) -> super::RouterSink
+    where
+        Self: Sized + 'static,
+    {
+        let (tx, rx) = channel(64);
+
+        let sink = LazyStreamingSink {
+            sink: Some((rx, self)),
+            inner: CompatSink::new(tx),
+        };
+
+        Box::new(sink)
+    }
+}
+
+pub struct LazyStreamingSink<T> {
+    sink: Option<(Receiver<Event>, T)>,
+    inner: CompatSink<Sender<Event>, Event>,
+}
+
+impl<T: StreamingSink> Sink for LazyStreamingSink<T> {
+    type SinkItem = Event;
+    type SinkError = ();
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        if let Some((rx, mut sink)) = self.sink.take() {
+            tokio02::spawn(async move {
+                if let Err(error) = sink.run(rx).await {
+                    error!(message = "Unexpected sink failure.", %error);
+                }
+            });
+        }
+
+        self.inner.start_send(item).map_err(drop)
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
+        self.inner.poll_complete().map_err(drop)
+    }
+}
```

Comment:
This could be dropped if we had access to an executor from `build` via `SinkContext`. This will just lazily spawn the bg task.

Reply:
I don't understand why we need to spawn a background task. There's only one task doing work here, right?

Reply:
Because we want to represent the actual logic of the sink as a its own task that gets fed items. Its just impossible to mix an async fn and Sink trait anyways. If this style works we may just want to feed a receiver to the sinks instead of using the sink trait. That would avoid the extra spawn but in end I don't think this hurts too much anyways.